### ISSUE TITLE
sysctl tunables

### ISSUE THREAD

**Issue by q53** [docker/docker#4717]
When launching a dockerised  process as root it is possible to execute some kind of init script to set sysctl values. But setting right ulimits is actually need  the -privileged flag.
When run as user (-u flag) is there a way to do subj? I did not find any in docs.
Setting of any other stuff (like /dev/shm size, https://github.com/dotcloud/docker/issues/2606) would be great.

**Comment by cpuguy83** [comment]
I feel like the shmmax/sysctl issue is separate. Can someone open a separate issue?

Also, full ulimit support is now implemented in #9437

**Comment by jgelens** [comment]
Great, now the ability for setting somaxconn and other similar sysctl vars

**Comment by pires** [comment]
+1

**Comment by mrbobbytables** [comment]
+1 here. This would solve several sticking issues I've had to work around.

**Comment by jeefy** [comment]
+1

**Comment by mcieslik-mctp** [comment]
+1 it almost a year...

**Comment by ppolv** [comment]
+1

**Comment by jagalan** [comment]
+1

**Comment by unclejack** [comment]
ulimits are now implemented by #9437. I've changed the title of this issue to reflect that it's only about sysctl tunables now. I'll reopen this so we don't have to create a new issue for container level sysctl tunables.

**Comment by jeremyeder** [comment]
Use case:  containerized java app or database performs better when using hugepages.  As of now we have 2 options:

* Allocate hugepages in the host, then launch a container with the app, and the app can correctly use hugepages.
* Use a privileged container to allocate hugepages, and then
 * Reuse that same privileged container to run the app, with all it's security implications...or, preferred:
 * Start a 2nd container for the app (non-privileged), and it can use hugepages allocated by the first.  They are not reclaimed by the kernel after the 1st container stops.
```
# grep -i HugePages_Total /proc/meminfo 
HugePages_Total:       0

# docker run --rm --privileged registry.access.redhat.com/rhel7/rhel-tools sysctl vm.nr_hugepages=10
vm.nr_hugepages = 10

# grep -i HugePages_Total /proc/meminfo 
HugePages_Total:      10

# docker run --rm --privileged registry.access.redhat.com/rhel7/rhel-tools sysctl vm.nr_hugepages=0
vm.nr_hugepages = 0

# grep -i HugePages_Total /proc/meminfo 
HugePages_Total:       0
```

**Comment by rhatdan** [comment]
@unclejack Is this something you plan on implementing?

docker run --sysctl vm.nr_hugepages=10 ...

**Comment by unclejack** [comment]
@rhatdan If it gets the required design approval, I'd be OK with implementing this. Just to make this clear, the design approval is only about the name of the flag and the other implementation details. This feature is really needed and there's no doubt it has to be implemented.

**Comment by LucaRav** [comment]
the thread started as a ulimit discussion/request but few asked and contributed with options and issues about kernel.shmmax and kernel.shmall. The default of 32MB is ridicolously small for databases unless you guys think Docker should just run web servers and scripting engines.
We need an option to 
1)
define it from the outset (# docker run kernel.<option>=val) (thus avoiding having to run silly scirpts) AND
2)
without --privileged

I really need at least 1 or databases in general will turn away from Docker tech.
thank you

**Comment by ewindisch** [comment]
Sysctls aren't namespaced and should not be part of a 'docker run'. It might be reasonable to have this as a scheduling hint or "requirement" such that a container will indicate it needs a minimum shmmax/shmall in order to run... but running the container shouldn't actually change that setting.

**Comment by q53** [comment]
>Sysctls aren't namespaced and should not be part of a 'docker run'
Excuse me?
[root@host ~]# sysctl -a | grep shm
kernel.shm_rmid_forced = 0
kernel.shmall = 268435456
kernel.shmmax = 4294967295
kernel.shmmni = 4096
[root@host ~]# lxc-attach -n test-1 /bin/su -
[root@test-1 ~]# sysctl -a | grep shm
kernel.shm_rmid_forced = 0
kernel.shmall = 4294967296
kernel.shmmax = 68719476736
kernel.shmmni = 4096
vm.hugetlb_shm_group = 0

Some of sysctls is very namespaced.

**Comment by ewindisch** [comment]
@q53 intresting. I did not know this. I'll confirm locally, but... interesting

**Comment by gdm85** [comment]
@LucaRav agree with you and looking forward for a way to consistently implement this.

**Comment by analytically** [comment]
+1

**Comment by christina57** [comment]
I need to install oracle db in container but stuck at "cannot set net.core.wmem_default". Please help!

**Comment by zrml** [comment]
Any news or comments on kernel shared mem params setting from the Docker developers?
Please see my previous post (see @zrml above)

I've read that apparently in kernel 3.16 a container can use the setting of the host.
Are you just waiting for that?
Is that what we want for every container by default? I'm not sure...

thanks

**Comment by zrml** [comment]
Hi,

Now that I'm getting into "auto-building" with Dockerfile(s) this security restriction is even worse because I cannot even install various products that need a minimum of 100MB shared mem segment (the default is a tiny 32MB on most distros).

This is a show stopper unless I could run the docker build (like unfortunately I have to do with the docker run command with the --privileged flag).
DO i really have to do these installs by hand? 

Thanks for any suggestion

**Comment by pior** [comment]
+1
Docker is basically useless for many of us dealing with web application until we can set  net.core.somaxconn (decently).
The same goes for hugepages...

**Comment by tianon** [comment]
There are a large number of sysctls that _are_ namespaced, most notably `net.*`.  Getting some kind of official solution for setting these is kind of a big deal.

**Comment by hadrien** [comment]
+1

**Comment by ewindisch** [comment]
@tianon having a good list of the sysctls to whitelist would be useful. I think it would be fine to just whitelist these in /proc.

**Comment by unclejack** [comment]
Having a list isn't sufficient. The namespaced tunables change from kernel to kernel, not from Docker to Docker release.

**Comment by ewindisch** [comment]
@unclejack without a mechanism to query from the kernel, a static whitelist is the only solution here... well, the only solution that isn't solved by user namespaces...

**Comment by ianludia** [comment]
+1

**Comment by sboichon** [comment]
+1

**Comment by kostko** [comment]
I would like to be able to run routing daemons in Docker containers. Some of these daemons require access to certain container-local network sysctls like the following:
```
/proc/sys/net/ipv6/conf/all/forwarding
/proc/sys/net/ipv6/conf/all/accept_redirects
```

Currently, the only way to allow this is to run the container in privileged mode, which is undesirable as `NET_ADMIN` capability should be enough. It would also be ok if one could specify `--privileged` but then remove all capabilities except `NET_ADMIN`. But this doesn't seem to work as all capabilities are always granted?

So I think that what should be implemented is a whitelist of container-local sysctls that should be made read-write or an option to make them so. Because it is not enough that these be set from the outside, sometimes one also wants the daemons running inside the containers to set them.

**Comment by rhatdan** [comment]
Have you tried using --cap-drop with --privileged? Not sure that works.

Something like --privileged --cap-drop=ALL --cap-add=NET_ADMIN

would be cool.

**Comment by kostko** [comment]
Yes, as I said above, unfortunately this doesn't work:

> It would also be ok if one could specify --privileged but then remove all capabilities except NET_ADMIN. But this doesn't seem to work as all capabilities are always granted?

From reading the Docker code it seems that once privileged mode is set, all capability settings are just ignored, so nothing is dropped :-(

**Comment by jeffbaier** [comment]
> Something like --privileged --cap-drop=ALL --cap-add=NET_ADMIN

I haven't tested this, but would the opposite work? Don't use the privileged flag, but instead --cap-add=ALL and then --cap-drop everything except what you need?

**Comment by kostko** [comment]
No, if you don't use the privileged flag then `/proc/sys` is mounted as read-only and you can't change anything.

**Comment by rhatdan** [comment]
So the solution is either a way to cause docker to not mount /proc/sys as read/only without specifying --priviliged or to make dropping caps work in --privileged mode.

Which might bring up a bigger problem in tha some of the sysctl settings are specific to namespaces, and mounting /proc and /proc/sys read/only might block other legitimate commands a confined application should be allowed to do.  This will be difficult to fix, without opening a whole to allow a container process to easily escape.

**Comment by deitch** [comment]
I noticed this hasn't been commented on in several months. What is the status of container-level tunables, specifically stuff like `net.core.*`? I am looking at `net.core.rmem_max` and `net.core.wmem_max`. Interestingly, these files do not even *exist* inside the container at `/proc/sys/net/core/`, making any kind of `sysctl` about them impossible. 

What is the recommended solution, then?

**Comment by jeremyeder** [comment]
In your specific case, you should adjust it on the host to a higher value.

This tunable is an example of the larger issue of per-namespace sysctls
that there seems to be some demand for (hence this issue).  The reality is
that the work needs to be done in the kernel first, and then exposed
through docker.  There are very few per-namespace sysctls right now.  The
network namespace has a handful and that's basically it.

On Fri, Aug 7, 2015 at 6:08 AM, Avi Deitcher <notifications@github.com>
wrote:

> I noticed this hasn't been commented on in several months. What is the
> status of container-level tunables, specifically stuff like net.core.*? I
> am looking at net.core.rmem_max and net.core.wmem_max. Interestingly,
> these files do not even *exist* inside the container at
> /proc/sys/net/core/, making any kind of sysctl about them impossible.
>
> What is the recommended solution, then?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/issues/4717#issuecomment-128662335>.
>



-- 

-- Jeremy Eder

**Comment by deitch** [comment]
Ah, so you are saying the kernel itself doesn't support these sysctls at the per-namespace level even, though, in theory, they should. Once the kernel does, then we can get down to providing the right support (read-write `/proc` etc.) at the tool level.

OK, thanks.

**Comment by faisyl** [comment]
We really need a fix for this. Ugly workarounds are breaking our automation workflow :(

**Comment by deitch** [comment]
@faisyl totally agree, but it sounds like it has to happen at the kernel level first.

**Comment by q53** [comment]
I'm terribly sorry, but what exactly kernel feature are you waiting for? Has it a feature request in the bugtracker?
IMHO, the list of required parameters (such as shm/net.core.* stuff) is a  very well known and pretty stable for a long time.

**Comment by deitch** [comment]
The ability to tune `net`-level parameters at a container level. If you want to change, e.g. `net.core.rmem_max` and `net.core.wmem_max` for a container, it has to happen at a host level, which kill any kind of container level automation.

**Comment by q53** [comment]
Ok, net.core.wmem_max is impossible to set from a container namespace. Does somebody in kernel team work on it? 
But still there are  lot of parameters which possible to set (net.ipv4.conf.* for example). Why not implement these things first?

**Comment by guilhermecvm** [comment]
+1 for somaxconn

**Comment by chiefy** [comment]
:+1: running into this myself.

**Comment by salanki** [comment]
+1

**Comment by pires** [comment]
Docker has had its _awesome_ moments when dealing with the community. Moron The Turtle is just another one of these! And the funny thing is I never recall commenting on this thread - but I was subscribed indeed.

Slow clap to the people who took the time to produce that piece of shit.

**Comment by jfrazelle** [comment]
You should check your attitude because we don't appreciate rude comments on threads. Read the contributing guidelines in the repo.

**Comment by tonylampada** [comment]
:+1: for smaxconn
(You can me to that list too :-))

**Comment by thaJeztah** [comment]
@tonylampada there's currently an open pull request here; https://github.com/docker/docker/pull/19265 but it uses a whitelist to only allow `sysctls` that are namespaced.

**Comment by tonylampada** [comment]
@thaJeztah interesting! Thanks for pointing that out.

**Comment by georgmzimmer** [comment]
+1 for somaxconn

**Comment by cpuguy83** [comment]
+1's do not help here.

If somaxconn is namespaced it can be added, if not, then it can't.
Based on the implementation in #19265, it looks like it is and will be supported.

**Comment by rhatdan** [comment]
If it is under net then it will be supported.

**Comment by cpuguy83** [comment]
Closed by the addition of `--sysctl` in #19265.

**Comment by nilyang** [comment]
I have the same question too, when use GDB ,add a `break point` ,then `run`, all ok but report an error:
```
warning: Error disabling address space randomization: Operation not permitted
```
When I google the reason ,I was found that this error can be resoved by change some kernel config like this:
```
root@7639183e92d0:~# sysctl -w kernel.randomize_va_space=0
sysctl: setting key "kernel.randomize_va_space": Read-only file system
```
But , as you see ,the **Read-only file system** error was reported agin, as `Operation not permitted`

**Comment by cpuguy83** [comment]
@nilyang you must run this on the host, not in the container... Or if you run it in the container you would have to use --privileged when creating the container.

**Comment by nilyang** [comment]
@cpuguy83 Thank you for your help ,but my problem has not been solved yet. I was send a command to my friend, and he's result was OK.
```
$ docker run --rm --privileged -ti alpine sh
```

**On my friend's MacOS**
![docker-privelges](https://cloud.githubusercontent.com/assets/7270458/17582829/463af9c6-5fe0-11e6-8ccb-7237b096490c.jpg)


**On My Windows 10**
```
 C:\Users\NilYang>docker run --rm --privileged -ti alpine sh
 / # echo 0> /proc/sys/kernel/randomize_va_space
 / # cat /proc/sys/kernel/randomize_va_space
 2
```

> As you can see , on **MacOS** the params runs OK, but on my **windows10** there was nothing happend!

 Then I guess that if the problem was just because my windows10? And my friend's OS is Mac.How ever, Thanks a lot!

**Comment by nilyang** [comment]
@cpuguy83 thanks a lot :+1: Btw, I found another question: this option can only take effect on **sh** ,but, when I start container with **bash** , there is no effect at all ~
Yes, my docker containers was runing on windows 10 , but I do not know why. Tody I send my question to my friend , who has a docker installed and runing  containers on MacOS , and he's result was OK  on both   `bash` and `sh` . After all ,I doubt that the only diffences about container  from my friend's and mine is `OS`.

**Comment by cpuguy83** [comment]
@nilyang It should work either way. The shell has no bearing in tweaking the kernel params.

**Comment by vingrad** [comment]
@cpuguy83 but it's impossible to use "privileged" with docker service. 
WARNING: unsupported parameter privileged

**Comment by Mishi-999** [comment]
Hii, i am facing a critical issue while running oracle rac database in docker container solely as net.core* parameters are not set in container. I also tried with --net=host but that does not seem to help as it uses host network which limit connectivity between two rac containers and also hinders the installation.

Is there any workaround for setting --sysctl net.core* parameters with docker run?
Thanks in advance.

**Comment by thaJeztah** [comment]
@Mishi-999 what workaround do you need? This feature is implemented in https://github.com/docker/docker/pull/19265, and part of docker 1.12

**Comment by Mishi-999** [comment]
Thanks for your response @thaJeztah , i am using Docker-version 1.12.1 on ubuntu 4.4.0-59-generic but running --sysctl net.core.rmem_default=262144 parameter in priviledged mode shows the error -
docker: Error response from daemon: oci runtime error: open /proc/sys/net/core/rmem_default: no such file or directory.
Parameters are set in /etc/sysctl.conf and yet on running sysctl -a i am unable to grep "net.core.rmem_default = 262144".
Can u please guide me in this issue.Sorry for bothering as i am new to docker.

**Comment by thaJeztah** [comment]
@Mishu-999 can you open a new issue with more details and the exact steps to reproduce?

**Comment by rhatdan** [comment]
This might mean that the net/core/rmem_default is not namespaced in the kernel so it can not be set per container.

**Comment by thaJeztah** [comment]
@rhatdan which means we should exclude that one from the list?

**Comment by justincormack** [comment]
@thaJeztah no, thats not helpful. We could maybe improve the error message.

**Comment by rhatdan** [comment]
@thaJeztah No, since we don't know what the list is?  :^(  This is a case where there is a kernel sysctl which is not namespaced, but currently we believe the all NET* sysctls are namespaced, and that is all we check.

**Comment by Mishi-999** [comment]
@thaJeztah created issue #30778

**Comment by daixiang0** [comment]
hi, i run inside a privileged container and do some setting with systctl:
```
bash-4.2# cat sysctl.conf
kernel.shmmax = 1000000000
kernel.shmmni = 4096
kernel.shmall = 4000000000
kernel.sem = 250 512000 100 2048
kernel.sysrq = 1
kernel.core_uses_pid = 1
kernel.msgmnb = 65536
kernel.msgmax = 65536
kernel.msgmni = 2048
net.ipv4.tcp_syncookies = 0
net.ipv4.conf.default.accept_source_route = 0
net.ipv4.tcp_tw_recycle = 1
net.ipv4.tcp_max_syn_backlog = 200000
net.ipv4.conf.all.arp_filter = 1
net.ipv4.ip_local_port_range = 32768 65535
net.core.netdev_max_backlog = 200000
vm.overcommit_memory = 2
fs.nr_open = 3000000
kernel.threads-max = 798720
kernel.pid_max = 798720
net.core.rmem_max = 2097152
net.core.wmem_max = 2097152

bash-4.2# sysctl -p sysctl.conf
kernel.shmmax = 1000000000
kernel.shmmni = 4096
kernel.shmall = 4000000000
kernel.sem = 250 512000 100 2048
kernel.sysrq = 1
kernel.core_uses_pid = 1
kernel.msgmnb = 65536
kernel.msgmax = 65536
kernel.msgmni = 2048
sysctl: cannot stat /proc/sys/net/ipv4/tcp_syncookies: No such file or directory
net.ipv4.conf.default.accept_source_route = 0
sysctl: cannot stat /proc/sys/net/ipv4/tcp_tw_recycle: No such file or directory
sysctl: cannot stat /proc/sys/net/ipv4/tcp_max_syn_backlog: No such file or directory
net.ipv4.conf.all.arp_filter = 1
net.ipv4.ip_local_port_range = 32768 65535
sysctl: cannot stat /proc/sys/net/core/netdev_max_backlog: No such file or directory
vm.overcommit_memory = 2
fs.nr_open = 3000000
kernel.threads-max = 798720
kernel.pid_max = 798720
sysctl: cannot stat /proc/sys/net/core/rmem_max: No such file or directory
sysctl: cannot stat /proc/sys/net/core/wmem_max: No such file or directory
```
Some proc file missed, i do not know how to fix.
here is my env:
```
[root@skydata_003 makeimage]# docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 28
Server Version: 17.06.2-ce
Storage Driver: devicemapper
 Pool Name: docker-253:1-1009778822-pool
 Pool Blocksize: 65.54kB
 Base Device Size: 53.69GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.042GB
 Data Space Total: 1.074TB
 Data Space Available: 475.8GB
 Metadata Space Used: 5.444MB
 Metadata Space Total: 2.147GB
 Metadata Space Available: 2.142GB
 Thin Pool Minimum Free Space: 10.74GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.140-RHEL7 (2017-05-03)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 31.26GiB
Name: skydata_003
ID: 3FMA:CIIC:S4WS:UKGO:PVTU:FENF:2TWV:HQZS:VMFE:6HLT:ICDP:52TW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 192.168.20.38
 127.0.0.0/8
Live Restore Enabled: false
```
```
[root@skydata_003 makeimage]# uname -a
Linux skydata_003 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
[root@skydata_003 makeimage]# cat /etc/redhat-release 
CentOS Linux release 7.3.1611 (Core) 
```

**Comment by thaJeztah** [comment]
@daixiang0 this issue is a feature request, and the feature was implemented last year. If you have a bug to report, please open a new issue with the information that's requested in the issue template that's shown when you open the issue. Make sure your system is fully up-to-date as CentOS uses a rolling update model (i.e. CentOS 7.3 reached end of life after 7.4 was released).

I'm locking this issue for comments to prevent it from collecting unrelated issues (which easily go overlooked on closed issues); if you arrive here because you want to report a bug; open a new issue instead, thanks!
